rfcs/53-consistent-hierarchy.md

```diff
@@ -0,0 +1,421 @@
+   issue, and all hiearchy traversals require heavy random access into memory.
+ - Updates are still not immediately visible within a stage and deferred until
+   the next command buffer flush.
+ - Hierarchy updates are now single threaded. Commands can be generated from
+   multiple systems at the same time, but they now require exclusive access to
+   apply.
+ - Each of the changed commands do a lot more work before and are signifgantly
+   more branch heavy. This could negatively impact the performance of
+   command-heavy workloads (i.e.  scene loading)
+ - The commands applied for controlling the hierarchy are computationally more
+   expensive, requiring more branching access and more `World::get` calls.
+ - Some potential implementations of this hierarchy requires normal ECS commands
+   to be aware of the hierarchy. Using `EntityCommands::remove` to remove
+   `Children` or `Parent` will break the invariants of the system without some
+   hooks to enforce the invariants on removal. This is partially mitigatable via
+   documentation until hooks or relations lands.
+
+## Rationale and alternatives
+The primary goal of this proposal is to get rid of the hierarchy maintenance
+system and the consistency issues that come with it's design, which it delivers
+on.
+
+### Alternative: Hierarchy as a Resource
+This would store the resource of the entire world as a dedicated Resource
+instead of as components in normal ECS storage. A prototypical design for this
+resource would be:
+
+```rust
+pub struct Hierarchy {
+  relations: Vec<(Option<usize>, Vec<usize>)>,
+  entities: Vec<Entity>,
+  indices: SparseArray<Entity, usize>,
+}
+```
+
+This would be strikingly similar to the `SparseSet` used in `bevy_ecs`. The main
+difference here is that the parallel `Vec`s are kept in topological order. When
+iterating over the hierarchy, children that are found before their parent will
+swapped with their parents.
+
+An alternative version of this hierarchy resource only stores parents. This may
+speed up iteration as individual relations are smaller during swaps, but finding
+children in the hierarchy is O(n) over the entire world. If the depth of each
+entity within the hierarchy, and the hierarchy is kept sorted in depth order this
+can be kept to O(n) in the next depth.
+
+#### Benefits
+
+ - Iteration in hierarchy order (parents first) is entirely linear. This may make
+   transform and other parent-reliant propagation signfigantly faster.
+ - Can be used with the command-driven interface proposed in this design.
+
+#### Drawbacks
+
+ - Merging and splitting hierarchies loaded from scenes can be quite difficult to
+   maintain.
+ - Iteration requires mutative access to the resource, which can limit
+   parallelism of systems traversing the hierarchy. The alternative requires a
+   dedicated O(n) sweep over the hierarchy whenever it's mutated.
+
+### Alternative: Linked List Hierarchy
```

Comment:
Linked list `Hierarchy` components are growing on me. Immediate, always-in-sync O(1) hierarchy updates are compelling. I've fixated on needing `Without<Parent>` queries in the past, but thats only important if we allow more than one root and/or treat "hierarchy membership / rootness" as a direct result of Hierarchy data. If we force explicit roots, this becomes a non-issue.

This creates a number of problems though:
1. How do we choose roots? Do we just spawn a single "canonical" root entity on our users' behalf? Do users denote a `Root` component on entities? Do we just ignore "hierarchy orphans" for things like transform propagation? Do we detect `parent == None` cases for entities without `Root` markers and warn?
2. Immediate hierarchy updates only work when both parent and child entities already have Hierarchy components. We would still need to rely on deferred parenting commands for the initial spawn of entities, or for entities that don't yet have the hierarchy component.
3. In light of (2), to actually make "immediate hierarchy updates" possible and worthwhile in practice, we would need to ensure that all common entities/bundles/prefabs intended to be hierarchy members have Hierarchy when they are spawned. This would mean a lot more "unnecessary" / empty hierarchy data in the ECS and more complicated bundles. Users defining their own bundles would need to add Hierarchy to get the "immediate" benefits. This feels brittle / in some cases users might not be sure if the parent is "ready", so they might need to try both (or just default to commands).
4. We would need to measure the extra `Query::get` child traversal perf relative to pre-computed child list iteration and decide if it is "worth it".

Reply:
After doing some work to parallelize transform propagation, I like this approach a lot less. Responses inline.

> How do we choose roots? Do we just spawn a single "canonical" root entity on our users' behalf? Do users denote a Root component on entities? Do we just ignore "hierarchy orphans" for things like transform propagation? Do we detect parent == None cases for entities without Root markers and warn?

A SparseSet marker component, maintained by the commands proposed here could work. We're already incurring heavy random read costs, so archetypal iteration at the root level likely isn't going to be much of a perf impact here.

I'd strongly recommend against a single root. Any propagation system we have now is parallelizing at the root level, and is reliant on efficiently identifying them and going wide splitting on the individual trees in the forest.

Other than the marker component, the only way I can think of is a resource root, which introduces more consistency problems.

> Immediate hierarchy updates only work when both parent and child entities already have Hierarchy components. We would still need to rely on deferred parenting commands for the initial spawn of entities, or for entities that don't yet have the hierarchy component.

The design here still calls for a command-based approach, so iteration is read-only. The `O(1)` nature just changes the costs of running those commands. Within those commands, we can update the state of the new child/sibling/parents by adding the `Hierarchy` component.

> In light of (2), to actually make "immediate hierarchy updates" possible and worthwhile in practice, we would need to ensure that all common entities/bundles/prefabs intended to be hierarchy members have Hierarchy when they are spawned. This would mean a lot more "unnecessary" / empty hierarchy data in the ECS and more complicated bundles. Users defining their own bundles would need to add Hierarchy to get the "immediate" benefits. This feels brittle / in some cases users might not be sure if the parent is "ready", so they might need to try both (or just default to commands).

Agreed here. This seems like a major downside. I'll add this to the RFC.

> We would need to measure the extra Query::get child traversal perf relative to pre-computed child list iteration and decide if it is "worth it".

This is the real kicker. `Query::get` has a lot of overhead when you're using it to iterate over a large portion of the `World`. The tight cache-locality of `Children` simplifies this by a long shot.